[#791] Support Hardhat v3

daenamkim opened the pull request:
Resolves #791
Resolves #212

**Summary**
- Made the Ignition deploy override follow Ignition’s execution decisions for MultiBaas syncing.
- Added `syncExisting` and `requireChainIdMatch` options, plus chain ID validation.
- Expanded sample contracts/modules (Series pattern) and updated sample scripts/docs.
- Added tests workflow for non-draft PRs with a 3-minute timeout.
- Fixed tests and added a minimal Ignition fixture module for CI stability.

**Testing**
- `npm run build`
- `npm test`

**Notes**
- The Ignition deploy override tests use `test/fixtures/ignition` to avoid resolving sample project dependencies during CI.

dbxe committed:
Initial commit for v3 rewrite

dbxe committed:
.gitignore ignition/deployments

dbxe committed:
fix network config file

daenamkim committed:
update lock files

daenamkim committed:
update config template

daenamkim committed:
fix hardhat config

daenamkim committed:
update package.json

daenamkim committed:
update gitignore

daenamkim committed:
Merge remote-tracking branch 'origin/main' into 791-hardhat-v3-rewrite

 Conflicts:
	.gitignore
	LICENSE
	README.md
	package-lock.json
	package.json
	sample/README.md
	sample/hardhat.config.ts
	sample/package-lock.json
	sample/package.json
	sample/tsconfig.json
	src/index.ts
	src/type-extensions.ts
	src/types.ts
	tsconfig.json

daenamkim committed:
restore eslint config

daenamkim committed:
fix eslint config

daenamkim committed:
add husky prepare

daenamkim committed:
lint failure demo

daenamkim committed:
update eslint command

daenamkim committed:
remove legacy files

daenamkim committed:
remove unnecessary config

daenamkim committed:
remove legacy files

shoenseiwaso commented on src/internal/tasks/ignition-deploy.ts:
General comment that we should add more comments to this file to help us navigate what it's doing.

shoenseiwaso commented on src/internal/multibaas/client.ts:
Similar note that we should add comments to this file, both about what it's doing, and the strategy that it takes and why.

shoenseiwaso commented on sample/package.json:
```diff
@@ -1,25 +1,26 @@
+    "deploy:metacoin": "HARDHAT_NETWORK=development npx hardhat ignition deploy ignition/modules/MetaCoinModule.ts",
+    "deploy:greeter": "HARDHAT_NETWORK=development npx hardhat ignition deploy ignition/modules/GreeterModule.ts",
+    "deploy:link:greeter": "HARDHAT_NETWORK=development npx hardhat ignition deploy ignition/modules/LinkedGreeterModule.ts",
+    "deploy:proxy:greeter": "HARDHAT_NETWORK=development npx hardhat ignition deploy ignition/modules/ProxiedGreeterModule.ts",
```
Given this is a development network, perhaps add `HARDHAT_IGNITION_CONFIRM_DEPLOYMENT=false` to these lines to skip the prompt?

dbxe replied:
To build on this and add additional context, the QA run commands we used are along the lines of 
```
HARDHAT_IGNITION_CONFIRM_DEPLOYMENT=1 HARDHAT_IGNITION_CONFIRM_RESET=1 npm run deploy:metacoin -- --reset
```

So it may be good to add those flags to the package.json

daenamkim replied:
https://github.com/curvegrid/hardhat-multibaas-plugin/pull/807/changes/1bda4f108239b04bbada3717938120eb950977e0

shoenseiwaso commented on sample/hardhat.config.ts:
```diff
@@ -1,76 +1,49 @@
+      chainType: "l1",
+      url: `${deploymentConfig.deploymentEndpoint}/web3/${deploymentConfig.web3Key}`,
+      chainId: deploymentConfig.ethChainID,
+      accounts: [deploymentConfig.deployerPrivateKey],
```
Adding this skips waiting for 5 blocks, where on non-HRE local dev networks with insta-mining, requires mining 5 more blocks manually.

Note that this change as-is works functionally, but will result in a linting error which will need to be resolved.

```suggestion
      accounts: [deploymentConfig.deployerPrivateKey],
      ignition: {
        requireConfirmations: 1,
      },
```

daenamkim replied:
https://github.com/curvegrid/hardhat-multibaas-plugin/pull/807/commits/6439fa177b1964f0e0907592ddbb38f04721e89c

https://hardhat.org/ignition/docs/reference/config#requiredconfirmations shows `{ ignition: requiredConfirmations }` looks at the top-level

shoenseiwaso replied:
Yes, but the point is that we want to wait for a single confirmation only on local development. On other chains, it might be advisable to wait for the default number of blocks (5?) to avoid issues with eventual consistency. WDYT?

daenamkim replied:
I have double checked `requiredConfirmations` option, but it did not work when I added it in to `development` but worked when I added it at top-level. The `ignition` option in `development` is ignored. 🤔 

```ts
  // === CASE 1: WORK
  // ignition: {
  //   requiredConfirmations: 1,
  // },
  networks: {
    development: {
      type: "http",
      chainType: "l1",
      url: `${deploymentConfig.deploymentEndpoint}/web3/${deploymentConfig.web3Key}`,
      chainId: deploymentConfig.ethChainID,
      accounts: [deploymentConfig.deployerPrivateKey],
      // === CASE 2: NOT WORK
      // ignition: {
      //   requiredConfirmations: 1,
      // },
    },
  },
```

### CASE 1: WORK

See the end of log

```sh
START=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
npm run deploy:greeter -- --reset
END=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
echo "start=${START} end=${END}"

> hardhat-multibaas-plugin-sample-v3@1.0.0 deploy:greeter
> npm run deploy:base -- ignition/modules/GreeterModule.ts --reset


> hardhat-multibaas-plugin-sample-v3@1.0.0 deploy:base
> HARDHAT_NETWORK=development HARDHAT_IGNITION_CONFIRM_DEPLOYMENT=false npx hardhat ignition deploy ignition/modules/GreeterModule.ts --reset

✔ Confirm reset of deployment "chain-1337" on chain 1337? … yes
Hardhat Ignition 🚀

Deploying [ GreeterModule ]

Batch #1
  Executed GreeterModule#Greeter

[ GreeterModule ] successfully deployed 🚀

Deployed Addresses

GreeterModule#Greeter - 0x79fdB736e52b591ef6272C04A0B6E8F690d5786F
MultiBaas: Contract "Greeter 1.0" already created. Skipping creation.
MultiBaas: Deleting old address 0x070C374dfd107986F95d53264659a2e2B119eCe3 with alias "greeter"
MultiBaas: Creating address 0x79fdB736e52b591ef6272C04A0B6E8F690d5786F with alias "greeter"
MultiBaas: Linking contract "greeter 1.0" to address "greeter"
start=2025-12-30T10:32:42Z end=2025-12-30T10:32:53Z ### ABOUT 10 seconds (1 block)
```

### CASE 2: NOT WORK

See the end of log

```sh
START=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
npm run deploy:greeter -- --reset
END=$(date -u +"%Y-%m-%dT%H:%M:%SZ")
echo "start=${START} end=${END}"

> hardhat-multibaas-plugin-sample-v3@1.0.0 deploy:greeter
> npm run deploy:base -- ignition/modules/GreeterModule.ts --reset


> hardhat-multibaas-plugin-sample-v3@1.0.0 deploy:base
> HARDHAT_NETWORK=development HARDHAT_IGNITION_CONFIRM_DEPLOYMENT=false npx hardhat ignition deploy ignition/modules/GreeterModule.ts --reset

✔ Confirm reset of deployment "chain-1337" on chain 1337? … yes
Hardhat Ignition 🚀

Deploying [ GreeterModule ]

Batch #1
  Executed GreeterModule#Greeter

[ GreeterModule ] successfully deployed 🚀

Deployed Addresses

GreeterModule#Greeter - 0x070C374dfd107986F95d53264659a2e2B119eCe3
MultiBaas: Contract "Greeter 1.0" already created. Skipping creation.
MultiBaas: Deleting old address 0xc6952ca5C0C4Dab0EC732959958356F55c8Ca358 with alias "greeter"
MultiBaas: Creating address 0x070C374dfd107986F95d53264659a2e2B119eCe3 with alias "greeter"
MultiBaas: Linking contract "greeter 1.0" to address "greeter"
start=2025-12-30T10:30:26Z end=2025-12-30T10:31:12Z ### ABOUT 50 seconds (5 blocks)
```

### SUGGESTION

What do you think of this?

```ts
  ignition: {
    requiredConfirmations: networkName === "development" ? 1  : 5,
  },
  networks: {
    development: {
      type: "http",
      chainType: "l1",
      url: `${deploymentConfig.deploymentEndpoint}/web3/${deploymentConfig.web3Key}`,
      chainId: deploymentConfig.ethChainID,
      accounts: [deploymentConfig.deployerPrivateKey],
    },
  },
```

shoenseiwaso replied:
Crazy...I could of sworn I had this working locally, but I reverted to pull in your changes to review. Now I can't get it working on local again. Sorry for sending you down a rabbit hole on this...let's just go with:

```
  ignition: {
    requiredConfirmations: 1, // CAUTION: suitable for development environments only with on-demand block production
  },
```

daenamkim replied:
Thank you for clarifying! Happy to go with it!

shoenseiwaso commented on src/internal/multibaas/client.ts:
```diff
@@ -0,0 +1,460 @@
+
+    return `0x${linkedBytecode}`;
+  }
+
+  private async _ensureAddress(
+    address: string,
+    contractLabel: string,
+    options: MultiBaasLinkOptions,
+  ): Promise<Address> {
+    const existingByAddress = await this._tryGetAddress(address);
+    if (existingByAddress !== undefined && existingByAddress.alias !== "") {
+      if (
+        options.addressAlias !== undefined &&
+        options.addressAlias !== existingByAddress.alias
+      ) {
+        throw new Error(
+          `MultiBaas: The address ${address} already exists under alias "${existingByAddress.alias}"`,
+        );
+      }
+
+      console.log(
+        `MultiBaas: Address ${address} already created as "${existingByAddress.alias}"`,
+      );
+      return existingByAddress;
+    }
```
The logic here (?) is slightly broken if the smart contract source code is modified and redeployed.

Updated smart contract:

<img width="720" height="534" alt="Image" src="https://github.com/user-attachments/assets/a1b5ad6c-5672-4cd1-9c56-dbc60c0ca44e" />

Hardhat deployment

```
$ npm run deploy:greeter

> hardhat-multibaas-plugin-sample-v3@1.0.0 deploy:greeter
> HARDHAT_NETWORK=development HARDHAT_IGNITION_CONFIRM_DEPLOYMENT=false npx hardhat ignition deploy ignition/modules/GreeterModule.ts


[ GreeterModule ] Nothing new to deploy based on previous execution stored in ./ignition/deployments/chain-1337

Deployed Addresses

LinkedGreeterModule#Greeter - 0x5F2E93a9DFc6498466692B2Eb0Ed4c7c6ec02D98
LinkedGreeterModule#LinkedGreeter - 0x5F2E93a9DFc6498466692B2Eb0Ed4c7c6ec02D98
GreeterModule#Greeter - 0x1A282828dbB61617aA967Cabd72c3f3A5c203331
MetaCoinModule#ConvertLib - 0x1eCFf743fC1149a113EA3df7463a4177f5C20735
MetaCoinModule#MetaCoin - 0x27B95dDEf613159616a934494e0B8b4f3E4028A0
MultiBaas: Delete old contract greeter 1.0 to deploy a new one
MultiBaas: Creating contract "greeter 1.0"
MultiBaas: Address 0x1A282828dbB61617aA967Cabd72c3f3A5c203331 already created as "greeter"
MultiBaas: Linking contract "greeter 1.0" to address "greeter"
```

MultiBaas UI view:

<img width="706" height="509" alt="Image" src="https://github.com/user-attachments/assets/343a2433-d784-4b13-afbe-b47ce1e35bd5" />

Basically, it detected that the contract bytecode changed, but it didn't link it to the new address.

daenamkim replied:
https://github.com/curvegrid/hardhat-multibaas-plugin/pull/807/commits/d44551546e75c3bff3f646558bed218a7bb9e1f2

I confirmed both the bytecode and addresses were updated.

shoenseiwaso replied:
It's desirable though that the plugin automatically detects what's changed on-chain, and then updates MultiBaas to reflect that. This is what the current plugin does. Basically:

1. If bytecode has changed, upload the new contract to MultiBaas and increment the version (if not specified).
2. If the deployment address has changed, relink the contract to the new address.

`--reset` basically forces Ignition to do everything again. Which is fine, but we still want the plugin to do the right thing if, for example, the bytecode has not changed, but Ignition has redeployed the contract, then MultiBaas should be updated to the new address.

Let me know if it's not clear! Happy to chat about it.

daenamkim replied:
I think [this change](https://github.com/curvegrid/hardhat-multibaas-plugin/pull/807/changes/754ec5104910bf9e1ea96ff4461d3c7e5c5ab8b8) might be the best for this issue because of the [reconciliation](https://hardhat.org/ignition/docs/explanations/execution#previous-state-and-reconciliation).

- Ignition matches futures by their ID (module name + future name + parameters hash)
- If the ID matches a previous successful execution in the journal, it reuses that deployment
- Changing the contract code alone doesn't change the future ID

I noticed is adding many deployment directories as deploying contracts many times. We can remove them in the plugin but it might be okay to leave as it is since they are very small.

<img width="329" height="453" alt="image" src="https://github.com/user-attachments/assets/20577d35-a91e-4161-b21d-41821465355f" />

shoenseiwaso reviewed (commented):
First pass, not bad!

dbxe commented on sample/hardhat.config.ts:
```diff
@@ -46,7 +46,4 @@ export default defineConfig({
   },
-  ignition: {
-    requiredConfirmations: 1,
-  },
```
Per @shoenseiwaso comment above, I think we should keep this

daenamkim replied:
Just FYI, https://github.com/curvegrid/hardhat-multibaas-plugin/pull/807#discussion_r2650382563

daenamkim committed:
update lock file

daenamkim committed:
add a legacy branch info

daenamkim committed:
add comments

daenamkim committed:
add requiredConfirmations and fix eslint error

daenamkim committed:
fix deploy error

daenamkim committed:
add HARDHAT_IGNITION_CONFIRM_DEPLOYMENT, HARDHAT_IGNITION_CONFIRM_RESET flags and update sample/readme.md

daenamkim committed:
fix redeploy did not work and update readme

daenamkim committed:
should not use --reset flag

daenamkim committed:
add requiredConfirmations config suggestion

daenamkim committed:
update requiredConfirmations option with comments

daenamkim committed:
add a solution to deploy a contract!

this commit will resolve the following root causes:

- Ignition matches futures by their ID (module name + future name + parameters hash)
- If the ID matches a previous successful execution in the journal, it reuses that deployment
- Changing the contract code alone doesn't change the future ID

daenamkim committed:
add tests in sample

daenamkim committed:
replace assert with expect

daenamkim committed:
add tests for plugin

daenamkim committed:
update eslint config

daenamkim committed:
fix eslint config and errors

daenamkim committed:
add string type

daenamkim committed:
fix resolveConfigurationVariable func signature

daenamkim committed:
update lock file

daenamkim committed:
add more commands

daenamkim committed:
optimize commands

daenamkim committed:
remove force deployment logic

daenamkim committed:
update readme with exmaples

daenamkim committed:
add more command details

daenamkim committed:
update readme

daenamkim committed:
client.ts to handle contract

daenamkim committed:
refactor config

daenamkim committed:
refactor ignition-deploy.ts

daenamkim committed:
refactor client.ts

shoenseiwaso committed:
logic and clarity

shoenseiwaso committed:
soothe linter

shoenseiwaso committed:
manual testing for more complex contracts

shoenseiwaso committed:
tests in CI

shoenseiwaso committed:
fix tests in CI

daenamkim committed:
fix dist path in package.json

daenamkim committed:
strict check for mbConfig

daenamkim committed:
refactro json strinfy null

daenamkim committed:
refactpr devDependencies in package.json for developers

shoenseiwaso committed:
update package-lock.json

shoenseiwaso committed:
sample package lock

daenamkim committed:
add more config test coverage

daenamkim committed:
add more ignition test coverage

daenamkim committed:
refactor config and add client test coverage

daenamkim committed:
fix readme

daenamkim committed:
fix example

daenamkim commented on sample/README.md:
````diff
@@ -1,40 +1,149 @@
+```
 
-## Testing/Deploying smart contracts
+```sh
+npm run wipe chain-1337 GreeterModule#Greeter
 
-To run tests in this sample project, first go the parent root project to build the `hardhat-multibaas-plugin` by running
+> hardhat-multibaas-plugin-sample-v3@1.0.0 wipe
+> HARDHAT_NETWORK=development npx hardhat ignition wipe chain-1337 GreeterModule#Greeter
 
-```shell
-cd ..
-npm install && npm run build
+GreeterModule#Greeter state has been cleared
````
@Copilot Fixed. Can you take a look again?

daenamkim committed:
clean up readme

daenamkim committed:
fix command exmaples

shoenseiwaso reviewed (approved):
LGTM!
